scipy/optimize/_isotonic.py

```diff
@@ -0,0 +1,134 @@
+        Case weights.
+    increasing : bool
+        If True, fit monotonic increasing, i.e. isotonic, regression.
+        If False, fit a monotonic decreasing, i.e. antitonic, regression.
+        Default is True.
+
+    Returns
+    -------
+    x : (N,) array_like
+        Isotonic regression solution, i.e. an increasing (or decreasing) array
+        of the same length than y.
+    wx : (B,) array_like
+        Sum of case weights for all blocks / pools B.
+    r : (B+1,) array_like
+        Array of indices with the start position of each block / pool B.
+        For the j-th block, all values of ``x[r[j]:r[j+1]]`` are the same.
+
+    Notes
+    -----
+    Given data :math:`y` and case weights :math:`w`, the isotonic regression
+    solves the following optimization problem:
+
+    .. math::
+
+        \operatorname{argmin}_{x_i} \sum_i w_i (y_i - x_i)^2 \quad
+        \text{subject to } x_i \leq x_j \text{ whenever } i \leq j \,.
+
+    For every input value :math:`y_i`, it generates an interpolated value
+    :math:`x_i` which are increasing. This is accomplished by the PAVA.
+    The solution consists of pools or blocks, i.e. neighboring elements of
+    :math:`x`, e.g. :math:`x_i` and :math:`x_{i+1}`, that all have the same
+    value.
+
+    Most interestingly, the solution stays the same if the squared loss is
+    replaced by the wide class of Bregman functions which are the unique
+    class of strictly consistent scoring functions for the mean, see [2]_
+    and references therein.
+
+    References
+    ----------
+    .. [1] Busing, F. M. T. A. (2022).
+           Monotone Regression: A Simple and Fast O(n) PAVA Implementation.
+           Journal of Statistical Software, Code Snippets, 102(1), 1-25.
+           :doi:`10.18637/jss.v102.c01`
+    .. [2] Jordan, A.I., Mühlemann, A. & Ziegel, J.F.
+           Characterizing the optimal solutions to the isotonic regression
+           problem for identifiable functionals.
+           Ann Inst Stat Math 74, 489-514 (2022).
+           :doi:`10.1007/s10463-021-00808-0`
+
+    Examples
+    --------
+    This example demonstrates that ``isotonic_regression`` really solves a constrained
+    optimization problem.
+    >>> from scipy.optimize import isotonic_regression, minimize
+    >>> y = [1.5, 1.0, 4.0, 6.0, 5.7, 5.0, 7.8, 9.0, 7.5, 9.5, 9.0]
+    >>> def objective(yhat, y):
+    ...     return np.sum((yhat - y)**2)
+    >>> def constraint(yhat, y):
+    ...     # This is for a monotonically increasing regression.
+    ...     return np.diff(yhat)
+    >>> result = minimize(objective, x0=y, args=(y,), constraints=
+            [{'type': 'ineq', 'fun': lambda x: constraint(x, y)}])
+    >>> result.x
+    array([1.25, 1.25, 4., 5.56666667, 5.56666667, 5.56666667, 7.8, 8.25, 8.25, 9.25,
+           9.25])
+    >>> yhat, _, _ = isotonic_regression(y)
+    >>> yhat
+    array([1.25, 1.25, 4., 5.56666667, 5.56666667, 5.56666667, 7.8, 8.25, 8.25, 9.25,
+           9.25])
+    """
```

Comment:
The example is great. (User hat on)Now that  I have an example, what is the benefit of using this function instead of copy-pasting the example? (user hat off)

Reply:
Good point: Mental load and scalability to larger data!

Reply:
https://github.com/scipy/scipy/pull/17722/commits/b511a17eab23ab99f35a8114b4bedd1b6228fc27